src/goto-instrument/contracts/contracts.cpp

```diff
@@ -564,24 +555,31 @@ bool code_contractst::apply_function_contract(
     insert_before_swap_and_advance(function_body, target, ensures_pair.second);
   }
 
+  // ASSIGNS clause should not refer to any quantified variables,
+  // and only refer to the common symbols to be replaced.
+  exprt targets;
+  for(auto &target : assigns)
+    targets.add_to_operands(std::move(target));
+  common_replace(targets);
+
   // Create a series of non-deterministic assignments to havoc the variables
   // in the assigns clause.
-  if(assigns.is_not_nil())
+  if(!assigns.empty())
   {
-    assigns_clauset assigns_clause(assigns, log, ns);
+    assigns_clauset assigns_clause(targets.operands(), log, ns);
 
     // Retrieve assigns clause of the caller function if exists.
     auto caller_assigns =
       to_code_with_contract_type(ns.lookup(function).type).assigns();
 
-    if(caller_assigns.is_not_nil())
+    if(enforced_functions.find(function.c_str()) != enforced_functions.end())
```

Comment:
This seems to check whether the contracts for caller (parent in the call chain) are being being checked.

However, as we discussed (documented in the specs doc), we need to check subset when contracts are _any_ ancestor is being checked.

Reply:
+1

Reply:
Based on an offline discussions, we decided to remove all subset relationship checks. The order in which we instrument contract allow us to rely on contract replacement and inlining of sub-function calls to catch any possible illegal writes.